From 0d4365501a75d370d920a2b41c6bc006b15c01b7 Mon Sep 17 00:00:00 2001
From: Manish Verma <quic_maniverm@quicinc.com>
Date: Wed, 8 Jun 2022 18:47:35 +0530
Subject: [PATCH 234/500] bridge: Fix the bridge stats update function

1. For updating the stats, use the current core's stats pointer
2. Hold the exclusive access to the sequence counter while updating
   the bridge stats.

Change-Id: I16d33677ee402dc4436741b630dfb9eb17caced0
Signed-off-by: Manish Verma <quic_maniverm@quicinc.com>
---
 net/bridge/br_if.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/net/bridge/br_if.c b/net/bridge/br_if.c
index 467b8bcd9a39..06be0570ae3d 100644
--- a/net/bridge/br_if.c
+++ b/net/bridge/br_if.c
@@ -859,14 +859,16 @@ void br_dev_update_stats(struct net_device *dev,
 	if (!(dev->priv_flags & IFF_EBRIDGE))
 		return;
 
-	stats = per_cpu_ptr(dev->tstats, 0);
+	stats = this_cpu_ptr(dev->tstats);
 
+	local_bh_disable();
 	u64_stats_update_begin(&stats->syncp);
 	u64_stats_add(&stats->rx_packets, nlstats->rx_packets);
 	u64_stats_add(&stats->rx_bytes, nlstats->rx_bytes);
 	u64_stats_add(&stats->tx_packets, nlstats->tx_packets);
 	u64_stats_add(&stats->tx_bytes, nlstats->tx_bytes);
 	u64_stats_update_end(&stats->syncp);
+	local_bh_enable();
 }
 EXPORT_SYMBOL_GPL(br_dev_update_stats);
 
-- 
2.34.1

